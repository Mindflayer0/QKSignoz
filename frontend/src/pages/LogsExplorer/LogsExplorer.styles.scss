.logs-module-page {
	display: flex;
	.log-quick-filter-left-section {
		width: 0%;
		flex-shrink: 0;
	}

	.log-module-right-section {
		width: 100%;
		.log-explorer-query-container {
			display: flex;
			flex-direction: column;
			flex: 1;

			.logs-explorer-views {
				flex: 1;
				display: flex;
				flex-direction: column;
			}
		}
	}

	&.filter-visible {
		.log-quick-filter-left-section {
			width: 20%;
		}

		.log-module-right-section {
			width: 80%;
		}
	}
}
