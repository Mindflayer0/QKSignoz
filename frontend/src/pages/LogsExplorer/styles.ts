import { Col } from 'antd';
import styled from 'styled-components';

export const WrapperStyled = styled.div`
	color: rgba(255, 255, 255, 0.85);
`;

export const ButtonWrapperStyled = styled(Col)`
	margin-left: auto;
`;
