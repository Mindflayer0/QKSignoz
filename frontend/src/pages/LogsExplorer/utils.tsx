import {
	FiltersType,
	IQuickFiltersConfig,
} from 'components/QuickFilters/QuickFilters';
import { DataTypes } from 'types/api/queryBuilder/queryAutocompleteResponse';
import { Query } from 'types/api/queryBuilder/queryBuilderData';

export const prepareQueryWithDefaultTimestamp = (query: Query): Query => ({
	...query,
	builder: {
		...query.builder,
		queryData: query.builder.queryData?.map((item) => ({
			...item,
			orderBy: [{ columnName: 'timestamp', order: 'desc' }],
		})),
	},
});

// eslint-disable-next-line @typescript-eslint/naming-convention
export enum SELECTED_VIEWS {
	SEARCH = 'search',
	QUERY_BUILDER = 'query-builder',
	CLICKHOUSE = 'clickhouse',
}

export const LogsQuickFiltersConfig: IQuickFiltersConfig[] = [
	{
		type: FiltersType.CHECKBOX,
		attributeKey: {
			key: 'service.name',
			id: 'service.name--string--resource--true',
			dataType: DataTypes.String,
			type: 'resource',
			isColumn: true,
			isJSON: false,
		},
		defaultOpen: true,
	},
	// {
	// 	type: FiltersType.CHECKBOX,
	// 	attributeKey: {
	// 		key: 'host.name',
	// 		dataType: DataTypes.String,
	// 		type: 'resource',
	// 		isColumn: true,
	// 		isJSON: false,
	// 	},
	// 	defaultOpen: false,
	// },
	// {
	// 	type: FiltersType.CHECKBOX,
	// 	attributeKey: {
	// 		key: 'service.instance.id',
	// 		dataType: DataTypes.String,
	// 		type: 'resource',
	// 		isColumn: true,
	// 		isJSON: false,
	// 	},
	// 	defaultOpen: false,
	// },
];
