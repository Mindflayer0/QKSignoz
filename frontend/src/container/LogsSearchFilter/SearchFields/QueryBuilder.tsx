import { CloseOutlined } from '@ant-design/icons';
import { Button, Input } from 'antd';
import React, { useRef, useState } from 'react';
import { useHoverDirty } from 'react-use';

import FieldKey from './FieldKey';
import { CategoryHeading, QueryFieldContainer } from './styles';

function QueryField() {
	const fieldRef = useRef(null);
	const isHovering = useHoverDirty(fieldRef);
	return (
		<QueryFieldContainer
			ref={fieldRef}
			
		>
			<div style={{ flex: 3 }}>
				<FieldKey name="host_name" type="string" />
			</div>
			<div
				style={{
					flex: 1,
					display: 'flex',
					minWidth: 400,
					gap: '1rem',
					alignItems: 'center',
					justifyContent: 'center',
				}}
			>
				<Input style={{ flex: 1, minWidth: 100 }} />
				<Input style={{ flex: 'auto', minWidth: 200 }} />
				<Button
					icon={<CloseOutlined />}
					type="text"
					size="small"
					style={{
						visibility: isHovering ? 'visible' : 'hidden',
					}}
				/>
			</div>
		</QueryFieldContainer>
	);
}

function QueryBuilder() {
	return (
		<div>
			<CategoryHeading>QUERY BUILDER</CategoryHeading>

			<QueryField />
			<QueryField />
		</div>
	);
}

export default QueryBuilder;
