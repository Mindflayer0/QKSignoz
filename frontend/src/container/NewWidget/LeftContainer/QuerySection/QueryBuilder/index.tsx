import {
	DeleteOutlined,
	DownOutlined,
	EyeFilled,
	RightOutlined,
} from '@ant-design/icons';
import { Button, Col, Divider, Input, Row, Select, Spin, Tabs } from 'antd';
import { getMetricName } from 'api/metrics/getMetricName';
import MonacoEditor from 'components/Editor';
import React, { useCallback, useEffect, useMemo, useState } from 'react';
import { useSelector } from 'react-redux';
import { AppState } from 'store/reducers';
import AppReducer from 'types/reducer/app';
import { v4 as uuid } from 'uuid';

import { InputContainer, QueryBuilderWrapper, QueryWrapper } from '../styles';
import { TQueryCategories } from '../types';
import MetricsBuilder from './metricsBuilder';
import MetricTagKeyFilter from './MetricTagKeyFilter';
import { AggregateFunctions } from './Options';
import PromQLQueryBuilder from './promQL';

const { Option } = Select;
function QueryBuilder({
	name,
	onDelete,
	queryCategory,
	queryData,
	updateQueryData,
}: QueryBuilderProps): JSX.Element {
	const [metricName, setMetricName] = useState(
		queryData.queryBuilder.metricName,
	);
	const { isDarkMode } = useSelector<AppState, AppReducer>((state) => state.app);
	const [hideFromUI, setHideFromUI] = useState<boolean>(false);
	const handleHideFromUI = (): void => {
		setHideFromUI(!hideFromUI);
	};
	const handleQueryBuilderChange = ({
		aggregateFunction,
		metricName,
		tagFilters,
	}) => {
		if (aggregateFunction) {
			queryData.queryBuilder.aggregateOperator = aggregateFunction;
		}

		if (metricName) {
			queryData.queryBuilder.metricName = metricName;
		}

		if (tagFilters) {
			queryData.queryBuilder.tagFilters.items = tagFilters;
		}

		updateQueryData(queryData);
	};
	const handleClickhouseQueryChange = (clickHouseQuery) => {
		if (clickHouseQuery !== null) {
			queryData.clickHouseQuery = clickHouseQuery;
		}
		updateQueryData(queryData);
	};
	const handlePromQLQueryChange = ({ query, legend }) => {
		if (query) queryData.promQL.query = query;
		if (legend) queryData.promQL.legend = legend;

		updateQueryData(queryData);
	};
	const [metricNameList, setMetricNameList] = useState([]);
	const [metricNameLoading, setMetricNameLoading] = useState(false);
	const handleMetricNameSearch = async (searchQuery = '') => {
		setMetricNameList([]);
		setMetricNameLoading(true);
		const { payload } = await getMetricName(searchQuery);
		setMetricNameLoading(false);
		if (!payload || !payload.data) {
			return;
		}
		setMetricNameList(payload.data);
	};

	return (
		<QueryWrapper>
			<Row style={{ justifyContent: 'space-between' }}>
				<Row>
					<Button type="ghost" icon={<EyeFilled />}>
						{name}
					</Button>
					<Button
						type="ghost"
						icon={hideFromUI ? <RightOutlined /> : <DownOutlined />}
						onClick={handleHideFromUI}
					/>
				</Row>

				<Button type="ghost" danger icon={<DeleteOutlined />} onClick={onDelete} />
			</Row>
			{!hideFromUI && (
				<QueryBuilderWrapper isDarkMode={isDarkMode}>
					{queryCategory === 0 ? (
						<MetricsBuilder
							queryData={queryData}
							updateQueryData={updateQueryData}
							metricName={metricName}
						/>
					) : queryCategory === 1 ? (
						<MonacoEditor
							language="sql"
							theme="vs-dark"
							height="200px"
							onChange={handleClickhouseQueryChange}
							value={queryData.clickHouseQuery}
							// options={options}
						/>
					) : queryCategory === 2 ? (
						<PromQLQueryBuilder
							query={queryData.promQL.query}
							onQueryChange={(value) => handlePromQLQueryChange({ query: value })}
							legend={queryData.promQL.legend}
							onLegendChange={(value) => handlePromQLQueryChange({ legend: value })}
						/>
					) : null}
				</QueryBuilderWrapper>
			)}
		</QueryWrapper>
	);
}

interface QueryBuilderProps {
	queryCategory: 0 | 1 | 2;
}
export default QueryBuilder;
