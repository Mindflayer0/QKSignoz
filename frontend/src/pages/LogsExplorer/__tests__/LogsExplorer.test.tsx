import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { QueryBuilderProvider } from 'providers/QueryBuilder';
import MockQueryClientProvider from 'providers/test/MockQueryClientProvider';
import { I18nextProvider } from 'react-i18next';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import i18n from 'ReactI18';
import store from 'store';

import LogsExplorer from '..';

// mocking the graph components in this test as this should be handled separately
jest.mock(
	'container/TimeSeriesView/TimeSeriesView',
	() =>
		// eslint-disable-next-line func-names, @typescript-eslint/explicit-function-return-type, react/display-name
		function () {
			return <div>Time Series Chart</div>;
		},
);
jest.mock(
	'container/LogsExplorerChart',
	() =>
		// eslint-disable-next-line func-names, @typescript-eslint/explicit-function-return-type, react/display-name
		function () {
			return <div>Histogram Chart</div>;
		},
);

jest.mock('constants/panelTypes', () => ({
	AVAILABLE_EXPORT_PANEL_TYPES: ['graph', 'table'],
}));

jest.mock('d3-interpolate', () => ({
	interpolate: jest.fn(),
}));

describe('Logs Explorer Tests', () => {
	test('Logs Explorer default view test without data', async () => {
		const {
			getByText,
			getByRole,
			queryByText,
			getByTestId,
			queryByTestId,
		} = render(
			<MemoryRouter initialEntries={['/logs/logs-explorer']}>
				<Provider store={store}>
					<I18nextProvider i18n={i18n}>
						<MockQueryClientProvider>
							<QueryBuilderProvider>
								<LogsExplorer />
							</QueryBuilderProvider>
						</MockQueryClientProvider>
					</I18nextProvider>
				</Provider>
			</MemoryRouter>,
		);

		// check the presence of histogram chart
		expect(getByText('Histogram Chart')).toBeInTheDocument();

		// toggle the chart and check it gets removed from the DOM
		const histogramToggle = getByRole('switch');
		await userEvent.click(histogramToggle);
		expect(queryByText('Histogram Chart')).not.toBeInTheDocument();

		// check the presence of search bar and query builder and absence of clickhouse
		const searchView = getByTestId('search-view');
		expect(searchView).toBeInTheDocument();
		const queryBuilderView = getByTestId('query-builder-view');
		expect(queryBuilderView).toBeInTheDocument();
		const clickhouseView = queryByTestId('clickhouse-view');
		expect(clickhouseView).not.toBeInTheDocument();

		// check the presence of List View / Time Series View / Table View
		const listView = getByTestId('logs-list-view');
		const timeSeriesView = getByTestId('time-series-view');
		const tableView = getByTestId('table-view');
		expect(listView).toBeInTheDocument();
		expect(timeSeriesView).toBeInTheDocument();
		expect(tableView).toBeInTheDocument();

		// check the presence of old logs explorer CTA
		const oldLogsCTA = getByText('Switch to Old Logs Explorer');
		expect(oldLogsCTA).toBeInTheDocument();
	});
});
