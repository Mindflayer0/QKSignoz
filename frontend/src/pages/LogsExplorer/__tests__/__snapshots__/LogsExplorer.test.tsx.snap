// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Logs Explorer Tests Logs Explorer Page should render with data 1`] = `
.c6 {
  width: 100% !important;
  margin-bottom: 0.3rem;
  cursor: pointer;
}

.c6 .ant-card-body {
  padding: 0.3rem 0.6rem;
}

.c9.c9.c9 {
  min-width: 1.5rem;
  white-space: nowrap;
}

.c8 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  overflow: hidden;
  width: 100%;
}

.c7 {
  margin-left: 0.5rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 6px;
}

.c10 {
  display: inline-block;
  text-overflow: ellipsis;
  overflow: hidden;
  white-space: nowrap;
}

.c5 {
  height: 40rem !important;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 100%;
}

.c2.c2.c2 {
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
}

.c1.c1.c1 {
  text-align: right;
}

.c0 {
  padding-right: 8px;
  visibility: visible;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c4 {
  color: #ffffffd9;
}

<div>
  <div
    class="toolbar"
  >
    <div
      class="leftActions"
    >
      <div
        class="left-toolbar"
      >
        <div
          class="left-toolbar-query-actions"
        >
          <button
            class="ant-btn css-dev-only-do-not-override-2i2tap ant-btn-default search action-btn active-tab"
            type="button"
          >
            <svg
              class="lucide lucide-mouse-pointer-square"
              data-testid="search-view"
              fill="none"
              height="14"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="14"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M21 11V5a2 2 0 0 0-2-2H5a2 2 0 0 0-2 2v14a2 2 0 0 0 2 2h6"
              />
              <path
                d="m12 12 4 10 1.7-4.3L22 16Z"
              />
            </svg>
          </button>
          <button
            class="ant-btn css-dev-only-do-not-override-2i2tap ant-btn-default query-builder action-btn"
            type="button"
          >
            <svg
              class="lucide lucide-atom"
              data-testid="query-builder-view"
              fill="none"
              height="14"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="14"
              xmlns="http://www.w3.org/2000/svg"
            >
              <circle
                cx="12"
                cy="12"
                r="1"
              />
              <path
                d="M20.2 20.2c2.04-2.03.02-7.36-4.5-11.9-4.54-4.52-9.87-6.54-11.9-4.5-2.04 2.03-.02 7.36 4.5 11.9 4.54 4.52 9.87 6.54 11.9 4.5Z"
              />
              <path
                d="M15.7 15.7c4.52-4.54 6.54-9.87 4.5-11.9-2.03-2.04-7.36-.02-11.9 4.5-4.52 4.54-6.54 9.87-4.5 11.9 2.03 2.04 7.36.02 11.9-4.5Z"
              />
            </svg>
          </button>
        </div>
        <div
          class="histogram-view-controller"
        >
          <article
            class="ant-typography css-dev-only-do-not-override-2i2tap"
          >
            Histogram
          </article>
          <button
            aria-checked="true"
            class="ant-switch ant-switch-small css-dev-only-do-not-override-2i2tap ant-switch-checked"
            role="switch"
            type="button"
          >
            <div
              class="ant-switch-handle"
            />
            <span
              class="ant-switch-inner"
            >
              <span
                class="ant-switch-inner-checked"
              />
              <span
                class="ant-switch-inner-unchecked"
              />
            </span>
          </button>
        </div>
      </div>
    </div>
    <div
      class="timeRange"
    >
      <button
        class="ant-btn css-dev-only-do-not-override-2i2tap ant-btn-primary ant-btn-dangerous"
        data-testid="newExplorerCTA"
        type="button"
      >
        <span
          class="ant-btn-icon"
        >
          <span
            aria-label="compass"
            class="anticon anticon-compass"
            role="img"
          >
            <svg
              aria-hidden="true"
              data-icon="compass"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm0 820c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372zm198.4-588.1a32 32 0 00-24.5.5L414.9 415 296.4 686c-3.6 8.2-3.6 17.5 0 25.7 3.4 7.8 9.7 13.9 17.7 17 3.8 1.5 7.7 2.2 11.7 2.2 4.4 0 8.7-.9 12.8-2.7l271-118.6 118.5-271a32.06 32.06 0 00-17.7-42.7zM576.8 534.4l26.2 26.2-42.4 42.4-26.2-26.2L380 644.4 447.5 490 422 464.4l42.4-42.4 25.5 25.5L644.4 380l-67.6 154.4zM464.4 422L422 464.4l25.5 25.6 86.9 86.8 26.2 26.2 42.4-42.4-26.2-26.2-86.8-86.9z"
              />
            </svg>
          </span>
        </span>
        <span>
          Switch to Old Logs Explorer
        </span>
      </button>
      <div
        class="date-time-selector"
      >
        <div
          class="c0"
        >
          <article
            class="ant-typography c1 css-dev-only-do-not-override-2i2tap"
          />
        </div>
        <form
          class="ant-form ant-form-inline css-dev-only-do-not-override-2i2tap c2"
        >
          <div
            class="c3"
          >
            <div
              class="custom-time-picker"
            >
              <span
                class="ant-input-affix-wrapper timeSelection-input timeSelection-input-container css-dev-only-do-not-override-2i2tap"
              >
                <span
                  class="ant-input-prefix"
                >
                  <svg
                    class="lucide lucide-clock"
                    fill="none"
                    height="14"
                    stroke="currentColor"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="2"
                    viewBox="0 0 24 24"
                    width="14"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <circle
                      cx="12"
                      cy="12"
                      r="10"
                    />
                    <polyline
                      points="12 6 12 12 16 14"
                    />
                  </svg>
                </span>
                <input
                  class="ant-input css-dev-only-do-not-override-2i2tap"
                  placeholder="Last 30 min"
                  type="text"
                  value=""
                />
                <span
                  class="ant-input-suffix"
                >
                  <svg
                    class="lucide lucide-chevron-down"
                    fill="none"
                    height="14"
                    stroke="currentColor"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="2"
                    viewBox="0 0 24 24"
                    width="14"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="m6 9 6 6 6-6"
                    />
                  </svg>
                </span>
              </span>
            </div>
          </div>
        </form>
      </div>
    </div>
    <div
      class="rightActions"
    >
      <div>
        <button
          class="ant-btn css-dev-only-do-not-override-2i2tap ant-btn-primary right-toolbar"
          type="button"
        >
          <span
            class="ant-btn-icon"
          >
            <svg
              class="lucide lucide-play"
              fill="none"
              height="14"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="14"
              xmlns="http://www.w3.org/2000/svg"
            >
              <polygon
                points="5 3 19 12 5 21 5 3"
              />
            </svg>
          </span>
          <span>
            Stage & Run Query
          </span>
        </button>
      </div>
    </div>
  </div>
  <div
    class="c4"
  >
    <div
      class="ant-row css-dev-only-do-not-override-2i2tap"
      style="row-gap: 0;"
    >
      <div
        class="ant-col ant-col-xs-24 css-dev-only-do-not-override-2i2tap"
      >
        <div>
          <div
            class="qb-search-view-container"
          >
            <div
              style="position: relative;"
            >
              <div
                class="ant-select query-builder-search css-dev-only-do-not-override-2i2tap ant-select-multiple ant-select-show-arrow ant-select-show-search"
                style="width: 100%; min-width: 7.7rem;"
              >
                <div
                  class="ant-select-selector"
                >
                  <div
                    class="ant-select-selection-overflow"
                  >
                    <div
                      class="ant-select-selection-overflow-item ant-select-selection-overflow-item-suffix"
                      style="opacity: 1;"
                    >
                      <div
                        class="ant-select-selection-search"
                        style="width: 0px;"
                      >
                        <input
                          aria-autocomplete="list"
                          aria-controls="rc_select_TEST_OR_SSR_list"
                          aria-expanded="false"
                          aria-haspopup="listbox"
                          aria-owns="rc_select_TEST_OR_SSR_list"
                          autocomplete="off"
                          class="ant-select-selection-search-input"
                          id="rc_select_TEST_OR_SSR"
                          role="combobox"
                          type="search"
                          value=""
                        />
                        <span
                          aria-hidden="true"
                          class="ant-select-selection-search-mirror"
                        >
                           
                        </span>
                      </div>
                    </div>
                  </div>
                  <span
                    class="ant-select-selection-placeholder"
                  >
                    Search Filter : select options from suggested values, for IN/NOT IN operators - press "Enter" after selecting options
                  </span>
                </div>
                <span
                  aria-hidden="true"
                  class="ant-select-arrow"
                  style="user-select: none;"
                  unselectable="on"
                >
                  <span
                    aria-label="down"
                    class="anticon anticon-down ant-select-suffix"
                    role="img"
                  >
                    <svg
                      aria-hidden="true"
                      data-icon="down"
                      fill="currentColor"
                      focusable="false"
                      height="1em"
                      viewBox="64 64 896 896"
                      width="1em"
                    >
                      <path
                        d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
                      />
                    </svg>
                  </span>
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="ant-col ant-col-xs-24 css-dev-only-do-not-override-2i2tap"
      >
        <div
          class="logs-explorer-views-container"
        >
          <div>
            Histogram Chart
          </div>
          <div
            class="logs-explorer-views-types"
          >
            <div
              class="views-tabs-container"
            >
              <div
                class="ant-btn-group views-tabs css-dev-only-do-not-override-2i2tap"
              >
                <button
                  class="ant-btn css-dev-only-do-not-override-2i2tap ant-btn-default selected_view tab"
                  data-testid="logs-list-view"
                  type="button"
                  value="list"
                >
                  <span>
                    List view
                  </span>
                </button>
                <button
                  class="ant-btn css-dev-only-do-not-override-2i2tap ant-btn-default tab"
                  data-testid="time-series-view"
                  type="button"
                  value="graph"
                >
                  <span>
                    Time series
                  </span>
                </button>
                <button
                  class="ant-btn css-dev-only-do-not-override-2i2tap ant-btn-default tab"
                  data-testid="table-view"
                  type="button"
                  value="table"
                >
                  <span>
                    Table
                  </span>
                </button>
              </div>
              <div
                class="logs-actions-container"
              >
                <div
                  class="tab-options"
                >
                  <div
                    class="format-options-container"
                  >
                    <button
                      class="ant-btn css-dev-only-do-not-override-2i2tap ant-btn-default ant-btn-icon-only periscope-btn"
                      type="button"
                    >
                      <span
                        class="ant-btn-icon"
                      >
                        <svg
                          class="lucide lucide-sliders"
                          fill="none"
                          height="14"
                          stroke="currentColor"
                          stroke-linecap="round"
                          stroke-linejoin="round"
                          stroke-width="2"
                          viewBox="0 0 24 24"
                          width="14"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <line
                            x1="4"
                            x2="4"
                            y1="21"
                            y2="14"
                          />
                          <line
                            x1="4"
                            x2="4"
                            y1="10"
                            y2="3"
                          />
                          <line
                            x1="12"
                            x2="12"
                            y1="21"
                            y2="12"
                          />
                          <line
                            x1="12"
                            x2="12"
                            y1="8"
                            y2="3"
                          />
                          <line
                            x1="20"
                            x2="20"
                            y1="21"
                            y2="16"
                          />
                          <line
                            x1="20"
                            x2="20"
                            y1="12"
                            y2="3"
                          />
                          <line
                            x1="2"
                            x2="6"
                            y1="14"
                            y2="14"
                          />
                          <line
                            x1="10"
                            x2="14"
                            y1="8"
                            y2="8"
                          />
                          <line
                            x1="18"
                            x2="22"
                            y1="16"
                            y2="16"
                          />
                        </svg>
                      </span>
                    </button>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="logs-explorer-views-type-content"
            >
              <div
                class="logs-list-view-container"
              >
                <div
                  class="c5"
                  data-testid="logs-list-virtuoso"
                >
                  <div
                    class="ant-card ant-card-bordered css-dev-only-do-not-override-2i2tap"
                    style="width: 100%; margin-top: 20px;"
                  >
                    <div
                      class="ant-card-body"
                      style="padding: 0px; height: 100%;"
                    >
                      <div
                        data-test-id="virtuoso-scroller"
                        data-virtuoso-scroller="true"
                        style="height: 100%; outline: none; overflow-y: auto; position: relative;"
                        tabindex="0"
                      >
                        <div
                          data-viewport-type="element"
                          style="width: 100%; height: 100%; position: absolute; top: 0px;"
                        >
                          <div
                            data-test-id="virtuoso-item-list"
                            style="box-sizing: border-box; padding-top: 0px; padding-bottom: 0px; margin-top: 0px;"
                          >
                            <div
                              data-index="0"
                              data-item-index="0"
                              data-known-size="100"
                              item="[object Object]"
                            >
                              <div
                                class="ant-card ant-card-bordered c6 css-dev-only-do-not-override-2i2tap"
                              >
                                <div
                                  class="ant-card-body"
                                >
                                  <div
                                    class="log-line"
                                  >
                                    <div
                                      class="log-state-indicator"
                                    >
                                      <div
                                        class="line INFO"
                                      >
                                         
                                      </div>
                                    </div>
                                    <div>
                                      <div
                                        class="c7"
                                      >
                                        <div
                                          class="c8"
                                        >
                                          <span
                                            aria-label="Log : "
                                            class="ant-typography ant-typography-secondary ant-typography-ellipsis ant-typography-single-line c9 log-field-key css-dev-only-do-not-override-2i2tap"
                                          >
                                            Log : 
                                            <span
                                              aria-hidden="true"
                                              style="position: fixed; display: block; left: 0px; top: 0px; z-index: -9999; visibility: hidden; pointer-events: none; font-size: 0px; word-break: keep-all; white-space: nowrap;"
                                            >
                                              lg
                                            </span>
                                            <span
                                              aria-hidden="true"
                                              style="position: fixed; display: block; left: 0px; top: 0px; z-index: -9999; visibility: hidden; pointer-events: none; font-size: 0px; width: 0px; white-space: normal; margin: 0px; padding: 0px;"
                                            >
                                              <span
                                                aria-hidden="true"
                                              >
                                                ...
                                              </span>
                                            </span>
                                          </span>
                                          <div
                                            class="c10 log-value"
                                          >
                                            2024-02-15T21:20:22.035Z	INFO	frontend/best_eta.go:106	Dispatch successful	{"service": "frontend", "trace_id": "0e727d00a1560dc7", "span_id": "0e727d00a1560dc7", "driver": "T703638C", "eta": "2m0s"}
                                          </div>
                                        </div>
                                        <div
                                          class="c8"
                                        >
                                          <span
                                            aria-label="Timestamp : "
                                            class="ant-typography ant-typography-secondary ant-typography-ellipsis ant-typography-single-line c9 log-field-key css-dev-only-do-not-override-2i2tap"
                                          >
                                            Timestamp : 
                                            <span
                                              aria-hidden="true"
                                              style="position: fixed; display: block; left: 0px; top: 0px; z-index: -9999; visibility: hidden; pointer-events: none; font-size: 0px; word-break: keep-all; white-space: nowrap;"
                                            >
                                              lg
                                            </span>
                                            <span
                                              aria-hidden="true"
                                              style="position: fixed; display: block; left: 0px; top: 0px; z-index: -9999; visibility: hidden; pointer-events: none; font-size: 0px; width: 0px; white-space: normal; margin: 0px; padding: 0px;"
                                            >
                                              <span
                                                aria-hidden="true"
                                              >
                                                ...
                                              </span>
                                            </span>
                                          </span>
                                          <div
                                            class="c10 log-value"
                                          >
                                            2024-02-16T02:50:22+05:30
                                          </div>
                                        </div>
                                      </div>
                                    </div>
                                  </div>
                                </div>
                              </div>
                            </div>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="explorer-options"
            style="background: transparent;"
          >
            <div
              class="view-options"
            >
              <div
                class="ant-select views-dropdown css-dev-only-do-not-override-2i2tap ant-select-single ant-select-allow-clear ant-select-show-arrow ant-select-loading ant-select-show-search"
                style="min-width: 170px;"
              >
                <div
                  class="ant-select-selector"
                >
                  <span
                    class="ant-select-selection-search"
                  >
                    <input
                      aria-autocomplete="list"
                      aria-controls="rc_select_TEST_OR_SSR_list"
                      aria-expanded="false"
                      aria-haspopup="listbox"
                      aria-owns="rc_select_TEST_OR_SSR_list"
                      autocomplete="off"
                      class="ant-select-selection-search-input"
                      id="rc_select_TEST_OR_SSR"
                      role="combobox"
                      type="search"
                      value=""
                    />
                  </span>
                  <span
                    class="ant-select-selection-placeholder"
                  >
                    Select a view
                  </span>
                </div>
                <span
                  aria-hidden="true"
                  class="ant-select-arrow ant-select-arrow-loading"
                  style="user-select: none;"
                  unselectable="on"
                >
                  <span
                    aria-label="loading"
                    class="anticon anticon-loading anticon-spin"
                    role="img"
                  >
                    <svg
                      aria-hidden="true"
                      data-icon="loading"
                      fill="currentColor"
                      focusable="false"
                      height="1em"
                      viewBox="0 0 1024 1024"
                      width="1em"
                    >
                      <path
                        d="M988 548c-19.9 0-36-16.1-36-36 0-59.4-11.6-117-34.6-171.3a440.45 440.45 0 00-94.3-139.9 437.71 437.71 0 00-139.9-94.3C629 83.6 571.4 72 512 72c-19.9 0-36-16.1-36-36s16.1-36 36-36c69.1 0 136.2 13.5 199.3 40.3C772.3 66 827 103 874 150c47 47 83.9 101.8 109.7 162.7 26.7 63.1 40.2 130.2 40.2 199.3.1 19.9-16 36-35.9 36z"
                      />
                    </svg>
                  </span>
                </span>
              </div>
              <button
                class="ant-btn css-dev-only-do-not-override-2i2tap ant-btn-round ant-btn-default"
                disabled=""
                type="button"
              >
                <svg
                  class="lucide lucide-disc3"
                  fill="none"
                  height="16"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                  viewBox="0 0 24 24"
                  width="16"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <circle
                    cx="12"
                    cy="12"
                    r="10"
                  />
                  <path
                    d="M6 12c0-1.7.7-3.2 1.8-4.2"
                  />
                  <circle
                    cx="12"
                    cy="12"
                    r="2"
                  />
                  <path
                    d="M18 12c0 1.7-.7 3.2-1.8 4.2"
                  />
                </svg>
                <span>
                   Save this view
                </span>
              </button>
            </div>
            <hr />
            <div
              class="actions"
            >
              <button
                class="ant-btn css-dev-only-do-not-override-2i2tap ant-btn-circle ant-btn-default"
                type="button"
              >
                <svg
                  class="lucide lucide-concierge-bell"
                  fill="none"
                  height="16"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                  viewBox="0 0 24 24"
                  width="16"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M2 18a2 2 0 0 1 2-2h16a2 2 0 0 1 2 2v2H2v-2Z"
                  />
                  <path
                    d="M20 16a8 8 0 1 0-16 0"
                  />
                  <path
                    d="M12 4v4"
                  />
                  <path
                    d="M10 4h4"
                  />
                </svg>
              </button>
              <button
                class="ant-btn css-dev-only-do-not-override-2i2tap ant-btn-circle ant-btn-default"
                type="button"
              >
                <svg
                  class="lucide lucide-plus"
                  fill="none"
                  height="16"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                  viewBox="0 0 24 24"
                  width="16"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M5 12h14"
                  />
                  <path
                    d="M12 5v14"
                  />
                </svg>
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
